src/py2spack/loading.py

```diff
@@ -0,0 +1,266 @@
+
+    def _get(self, name: str) -> dict | APIError:
+        """Load info for the available distribution files from PyPI.
+
+        Available versions are cached.
+        """
+        r = requests.get(
+            f"https://pypi.org/simple/{name}/",
+            headers={"Accept": "application/vnd.pypi.simple.v1+json"},
+            timeout=10,
+        )
+        if r.status_code != 200:
+            if r.status_code == 404:
+                msg = f"Package {name} not found on PyPI (status code 404)"
+                return APIError(msg)
+
+            msg = (
+                f"Error when querying JSON API (status code {r.status_code})."
+                f" Response: {r.text}"
+            )
+            return APIError(msg)
+
+        data: dict = r.json()
+        versions = data["versions"]
+
+        # parse and sort versions
+        parsed_versions = sorted(
+            {vv for v in versions if (vv := _acceptable_version(v))}
+        )
+
+        self.version_cache[name] = parsed_versions
+
+        return data
+
+    def get_versions(self, name: str) -> List[vn.Version]:
+        """Get acceptable versions for package (from cache if possible)."""
+        if name not in self.version_cache:
+            data = self._get(name)
+
+            if isinstance(data, APIError):
+                self.version_cache[name] = []
+
+        return self.version_cache[name]
+
+    def get_files(
+        self, name: str, last_n_versions=-1
+    ) -> List[Dict[Any, Any]] | APIError:
+        """Get metadata for available distribution files from PyPI.
+
+        Returns:
+            - List of dictionaries including filename, download url, version,
+            sdist archive extension, and archive sha256 hash.
+            If an error occurs, returns APIError object.
+        """
+        data = self._get(name)
+        if isinstance(data, APIError):
+            return data
+        files = data["files"]
+
+        # for now we only support tarball archives like .tar.gz
+        files_known_format = [
+            f for f in files if _known_archive_format(f["filename"])
+        ]
+
+        if len(files_known_format) == 0:
+            msg = (
+                "No files with known archive format found (note: wheel file"
+                " parsing not supported)"
+            )
+            return APIError(msg)
+
+        # for each file, get the filename, url, version, extension, and sha256
+        # TODO: in case of an error, skip the file or return the error?
+        files_parsed: List[Dict[str, str]] = []
+        for f in files_known_format:
+            filename = f["filename"]
+            archive_ext = _parse_archive_extension(filename)
+            if isinstance(archive_ext, APIError):
+                continue
+
+            v = _parse_version(filename, name, archive_ext)
+
+            if v is None:
+                continue
+
+            try:
+                sha256 = f["hashes"]["sha256"]
```

Comment:
Can you explain to me when the "sha256" hash could be missing? Why is it correct to continue in that case?

Reply:
@egparedes While the `hashes` field is required, it can be that there is no hash in it, or another hash than `sha256`, according to https://peps.python.org/pep-0691/.
In that case I want to simply skip this specific version and try the next one.

Reply:
spack actually also supports other hashes than sha256, so I could in addition check for other hashes that spack supports